happi/audit.py

```diff
@@ -0,0 +1,218 @@
+                           container, device)
+        elif not container:
+            logger.warning('No container provided for %s', device)
+
+    def validate_args(self, item):
+        """
+        Validates the args of an item
+        """
+        [self.create_arg(item, arg) for arg in item.args]
+
+    def validate_kwargs(self, item):
+        """
+        Validates the kwargs of an item
+        """
+        dict((key, self.create_arg(item, val))
+             for key, val in item.kwargs.items())
+
+    def create_arg(self, item, arg):
+        """
+        Function borrowed from loader to create
+        correctly typed arguments from happi information
+        """
+        if not isinstance(arg, str):
+            return arg
+        try:
+            return fill_template(arg, item, enforce_type=True)
+        except Exception as e:
+            logger.warning('Probably provided invalid argument: %s for %s, %s',
+                           arg, item.name, e)
+
+    def validate_device_class(self, item):
+        """
+        Validates device_class field
+        """
+        device_class = item.get('device_class')
+        device = item.get('name')
+        if not device_class:
+            logger.warning('Detected a None vlaue for %s. '
+                           'The device_class cannot be None', device)
+        else:
+            try:
+                mod, cls = device_class.rsplit('.', 1)
+            except (Exception, ValueError) as e:
+                logger.warning('Wrong device name format: %s for %s, %s',
+                               device_class, device, e)
+            else:
+                try:
+                    import_class(device_class)
+                except ImportError as ex:
+                    logger.warning(ex)
+
+    def validate_extra_attributes(self, item):
+        """
+        Validate items that have extra attributes
+        """
+        attr_list = []
+        extr_list = ['creation', 'last_edit', '_id', 'type']
+
+        for (key, value), s in zip(item.items(), item.entry_info):
+            attr_list.append(s.key)
+
+        key_list = [key for key, value in item.items()]
+        diff = set(key_list) - set(attr_list) - set(extr_list)
+        if diff:
+            logger.warning('Device %s has extra attributes %s',
+                           item.name, diff)
+
+    def print_report_message(self, message):
+        logger.info('')
+        logger.info('--------- %s ---------', message)
+        logger.info('')
+
+    def parse_database(self, database_path):
+        """
+        Validates if an entry is a valid entry in the database
+
+        Parameters
+        ----------
+        database_path: str
+            Path to the database to be validated
+
+        """
+        client = Client(path=database_path)
+        items = client.all_items
+
+        self.print_report_message('VALIDATING ENTRIES')
+        client.validate()
```

Comment:
Also I might have to make my own `validate()` here - as i kind of started previously, but i am not sure, for now when i do this it gives me warning like these:
`Failed to validate cxi_dg1_slits because Missing mandatory information (device_class, location_group, functional_group) for Slits` - i might have to have extra checks here or something else, these devices seem to have a device_class already...

Reply:
Also that repo is super confusing right now: `master` branch is an old snapshot that should probably fail the audit, while the `deploy` branch is a new setup that should probably mostly pass.

Reply:
Yeah just realized that this morning :).
Is there a good way to differentiate between a module that we know it is valid but not in the python environment, other than specifically checking for `pcdsdevices`? - I want to be able to know when it is invalid/misspelled vs valid but not in the path?

Reply:
I'm not sure there are great tools for a generic python environment to check if a module is "real and missing" vs "not a real thing" (misspelled). Our best effort here could be to check pypi or conda, which should typically be available. I'm not sure if those have easily accessible python bindings or if we need to shell out e.g. `pip search pcdsdevices`. It also might be possible to get a list of installed modules and try to do a fuzzy match, e.g. "you specified the pcdsevices module, did you mean pcdsdevices?"

Reply:
Ohh that's a great idea, i should look into that, i did not think of checking with pypi and conda - i hope it is possible and not too bad to use :) - thank you for the suggestions here